doc/icainfo.1
.\" icainfo man page source
.\"
.\" use
.\"   groff -man -Tutf8 icainfo.1
.\" or
.\"   nroff -man icainfo.1
.\" to process this source
.\"
.TH ICAINFO 1 2022-11-21 IBM "icainfo user manual"
.SH NAME
icainfo \- print information about cryptographic functions supported by libica
.SH SYNOPSIS
.B icainfo
[-v | --version] [-h | --help] [-c | --list-curves] [-r | --list-rsa]
[-f | --list-fips-exceptions]
.SH DESCRIPTION
.B icainfo
prints a table that shows libica's support for various cryptographic
algorithms and information about FIPS support.

The icainfo output also indicates, whether it is in an error state.
Algorithms that are not FIPS approved are marked as blocked in all table
columns when running in FIPS mode. All algorithms are marked as blocked when
libica is in an error state.

Available hardware support is divided into two columns: dynamic hardware
means crypto cards, static hardware support means CPACF. Software support
is provided via openssl.

If a function is disabled via build option, this is indicated
by a dash '-' in the related column.

A shortened sample output is given below:
.P
.nf
      Cryptographic algorithm support
------------------------------------------------------
               |         hardware        |
 function      |   dynamic  |   static   |  software
---------------+------------+------------+------------
         SHA-1 |     no     |    yes     |    -
       SHA-224 |     no     |    yes     |    -
       SHA-256 |     no     |    yes     |    -
       SHA-384 |     no     |    yes     |    -
       SHA-512 |     no     |    yes     |    -
         GHASH |     no     |    yes     |    -
         P_RNG |  blocked   |  blocked   |  blocked
  DRBG-SHA-512 |     no     |    yes     |    -
        RSA ME |    yes     |     no     |    -
       RSA CRT |    yes     |     no     |    -
           ...
------------------------------------------------------
Built-in FIPS support: FIPS 140-3 mode active.
Software fallbacks are disabled.
.fi
.SH OPTIONS
.IP "-v or --version"
show libica version and copyright
.IP "-h or --help"
display this help and exit
.IP "-c or --list-curves"
show supported elliptic curves
.IP "-r or --list-rsa"
show min and max supported RSA key lengths
.IP "-f or --list-fips-exceptions"
show available algorithms that are non-approved
.SH RETURN VALUE
.IP 1
unknown or invalid argument on invocation
.IP 0
successful program execution
.SH "SEE ALSO"
.BR icastats (1)



ChangeLog
v4.1.1
 - [PATCH] Fix aes-xts multi-part operations
   [PATCH] Fix make dist
v4.1.0
 - [FEATURE] FIPS: make libica FIPS 140-3 compliant
   [FEATURE] New API function ica_ecdsa_sign_ex()
   [FEATURE] New icainfo output option -r
 - [PATCH] Various bug fixes
v4.0.3
 - [PATCH] Reduce the number of open file descriptors
 - [PATCH] Various bug fixes
v4.0.2
 - [PATCH] Various bug fixes
v4.0.1
 - [PATCH] Various bug fixes
 - [PATCH] Compute HMAC from installed library
v4.0.0
 - [UPDATE] NO_SW_FALLBACKS is now the default for libica.so
   [UPDATE] Removed deprecated API functions including tests
   [UPDATE] Introduced 'const' for some API function parameters
   [FEATURE] icastats: new parm -k to display detailed counters
v3.9.0
 - [FEATURE] Add support for OpenSSL 3.0
   [FEATURE] icainfo: new parm -c to display available EC curves
v3.8.0
 - [FEATURE] provide libica-cex module to satisfy special security requirements
   [FEATURE] FIPS: enforce the HMAC check
 - [UPDATE] exploit autoselect domain support for ECC
 - [UPDATE] FIPS: Add SHA3 kats to fips powerup tests
v3.7.0
 - [FEATURE] FIPS: Add HMAC based library integrity check
 - [PATCH] icainfo: bugfix for RSA and EC related info for software column.
 - [PATCH] FIPS: provide output iv in cbc-cs decrypt as required by FIPS tests
 - [PATCH] FIPS: Fix DES and TDES key length
 - [PATCH] icastats: Fix stats counter format
v3.6.1
 - [PATCH] Fix x25519 and x448 handling of non-canonical values
v3.6.0
 - [FEATURE] Add MSA9 CPACF support for Ed25519, Ed448, X25519 and X448
v3.5.0
 - [FEATURE] Add MSA9 CPACF support for ECDSA sign/verify
v3.4.0
 - [FEATURE] Add SHA-512/224 and SHA-512/256 support
v3.3.3
 - [PATCH] Various bug fixes
v3.3.2
 - [PATCH] Skip ECC tests if required HW is not available
 - [PATCH] Update spec file
v3.3.1
 - [PATCH] Fix configure.ac to honour CFLAGS
v3.3.0
 - [FEATURE] Add CEX supported elliptic-curve crypto interfaces
 - [FEATURE] Add SIMD supported multiple-precision arithmetic interfaces
 - [FEATURE] Add interface to enable/disable SW fallbacks
 - [FEATURE] Add 'make check' target, test-suite rework
v3.2.1
 - [FEATURE] Use z14 PRNO-TRNG to seed SHA512-DRBG.
 - [PATCH] Various bug fixes.
v3.2.0
 - [FEATURE] New AES-GCM interface.
 - [UPDATE] Add symbol versioning.
v3.1.1
 - [PATCH] Various bug fixes related to old and new AES-GCM implementations.
 - [UPDATE] Add SHA3 test cases. Improved and extended test suite.
v3.1.0
 - [FEATURE] Add KMA support for AES-GCM.
 - [FEATURE] Add SHA-3 support.
 - [PATCH] Reject RSA keys with invalid key-length.
 - [PATCH] Allow zero output length for ica_random_number_generate.
 - [PATCH] icastats: Correct owner of shared segment when root creates it.
v3.0.2
 - [PATCH] Fix locking callbacks for openSSL APIs.
v3.0.1
 - [PATCH] Fixed msa level detection on zEC/BC12 GA1 and predecessors.
v3.0.0
 - [FEATURE] Added FIPS mode.
 - [PATCH] Sanitized exported symbols.
 - [PATCH] Removed deprecated APIs. Marked some APIs as deprecated.
 - [PATCH] Adapted to OpenSSL v1.1.0.
 - [PATCH] RSA key generation is thread-save now.
v2.6.2
 - [PATCH] Performance improvements for des-ctr, 3des-ctr and aes-ctr
v2.6.1
 - [Patch] Fixed buffer overflow on random generation
 - [Patch] Fixed DRBG entropy input
 - [Patch] Fixed icastats test case
v2.6.0
 - [FEATURE] Added NIST compliant deterministic random bit generator (DRBG)
v2.5.0
 - [FEATURE] Added streaming support for Galois Counter Mode (AES-GCM)
 - [FEATURE] Allow RSA Key generation for any granularity (57..4096 bits)
 - [FEATURE] New RSA CRT Key check API (Convertion of non-compliant keys)
 - [PATCH] Replaced deprecated RSA_key_generate function
 - [PATCH] Fixed accounting for GCM, CCM and CMAC. Added accounting for GHASH
 - [PATCH] Improved configure script to automatically build test cases
 - [UPDATE] Added SHA test cases. Improved and extended test suite
 v2.4.0
 Reworked the statistic measurement facility (icastats)
 - Statistics are now collected independent from the process context
 - Statistic data gathered per user
 - User based statistic management (reset/delete/summerize)
v2.3.0
 Changes since Version 2.2.0:
 - Introduced new API to recieve list of supported algorithms
v2.2.0
 Introduce support for more MSA4 extensions.
 New symetric encryption modes:
 - CBC-CS (DES, TDES, AES-[128|192|256])
 - CMAC (DES, TDES, AES-[128|192|256])
 - CCM (AES-[128|192|256])
 - GCM (AES-[128|256])
v2.1.1
 Changes since Version 2.1.0:
 - [PATCH] synchronize shared memory ref counting
 - [PATCH] Fix result/error handling in testcase for ica_get_version()
 - [PATCH] Fix temporary buffer allocation in ica_get_version()
 - [UPDATE] Interface (API) description
 - [PATCH] Fix commandline argument checking in RSA testcase
 - [PATCH] Remove duplicate entry from target list
v2.1.0
- Introduce support for MSA4 extensions. New symmetric encryption modes:
  CFB, CTR, OFB, XTS, CMAC
- New API calls:
  - ica_get_version(...)
v2.0.6
Fix for tracker item 3202845.
  Improve online help for icastats.
v2.0.5
Fix for performance improvement
  Remove signal handler wrapping of crypto instructions.
v2.0.4
Fix for STCK buffer length
  Correctly adjust the STCK buffer length in the PRNG based on whether
  we are running in 31 or 64 bit.
v2.0.3
minor wording fix in icastats tool:
  changed output wording from SHAxxx to SHA-xxx
Bugfix for:
  libica abends with illegal instuction when CPACF is disabled.
v2.0.2
Bug fixes in asm, stcke for 31 bit, and fixes in programming samples
v2.0.1
Add locking for shared memory segment, which is needed when icastats
is linked dynamically against libica.
v2.0
Substitute software fallbacks with lowlevel OpenSSL function calls.
Rewrite most of libica.
Only Linux on System z code.


src/icainfo.c
/* This program is released under the Common Public License V1.0
 *
 * You should have received a copy of Common Public License V1.0 along with
 * with this program.
 */

/*
 * Display a list of all CP Assist for Cryptographic Function (CPACF)
 * operations supported by libica on a system.
 *
 * Author(s): Ralph Wuerthner <rwuerthn@de.ibm.com>
 * 	      Holger Dengler <hd@linux.vnet.ibm.com>
 * 	      Benedikt Klotz <benedikt.klotz@de.ibm.com>
 * 	      Ingo Tuchscherer <ingo.tuchscherer@de.ibm.com>
 *
 * Copyright IBM Corp. 2007, 2011, 2014
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <getopt.h>
#include <libgen.h>
#include <unistd.h>
#include <dirent.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <errno.h>

#include <openssl/objects.h>

#include "fips.h"
#include "ica_api.h"
#include "s390_crypto.h"
#include "s390_ecc.h"

#if defined(NO_SW_FALLBACKS) && defined(NO_CPACF)
#define CMD_NAME "icainfo-cex"
#else
#define CMD_NAME "icainfo"
#endif
#define COPYRIGHT "Copyright IBM Corp. 2007, 2022."

#define CELL_SIZE 3

typedef struct {
	unsigned int nid;
	unsigned char *sname;
	unsigned int flags;
} s390_supported_curves_t;

static const unsigned int cca_nids[] = {
	NID_X9_62_prime192v1, NID_secp224r1, NID_X9_62_prime256v1,
	NID_secp384r1, NID_secp521r1, NID_brainpoolP160r1, NID_brainpoolP192r1,
	NID_brainpoolP224r1, NID_brainpoolP256r1, NID_brainpoolP320r1,
	NID_brainpoolP384r1, NID_brainpoolP512r1 };
static const unsigned int cca_nids_len = sizeof(cca_nids) / sizeof(unsigned int);

static const unsigned int cpacf_nids[] = {
	NID_X9_62_prime256v1, NID_secp384r1, NID_secp521r1,
	NID_ED25519, NID_ED448, NID_X25519, NID_X448 };
static const unsigned int cpacf_nids_len = sizeof(cpacf_nids) / sizeof(unsigned int);

#ifdef ICA_FIPS
static const unsigned int cpacf_fips_nids[] = {
	NID_X9_62_prime256v1, NID_secp384r1, NID_secp521r1 };
static const unsigned int cpacf_fips_nids_len = sizeof(cpacf_fips_nids) / sizeof(unsigned int);
#endif

static int is_msa9(void)
{
	unsigned int mech_len, j;
	libica_func_list_element *pmech_list = NULL;

	if (ica_get_functionlist(NULL, &mech_len) != 0) {
		perror("get_functionlist: ");
		return 0;
	}

	pmech_list = malloc(sizeof(libica_func_list_element)*mech_len);
	if (ica_get_functionlist(pmech_list, &mech_len) != 0) {
		perror("get_functionlist: ");
		free(pmech_list);
		return 0;
	}

	for (j = 0; j < mech_len; j++) {
		if (pmech_list[j].mech_mode_id == EC_DSA_SIGN) {
			if (pmech_list[j].flags & ICA_FLAG_SHW) {
				free(pmech_list);
				return 1;
			}
		}
	}

	free(pmech_list);
	return 0;
}

static int online_cca_card(void)
{
	unsigned int mech_len, j;
	libica_func_list_element *pmech_list = NULL;

	if (ica_get_functionlist(NULL, &mech_len) != 0) {
		perror("get_functionlist: ");
		return 0;
	}

	pmech_list = malloc(sizeof(libica_func_list_element)*mech_len);
	if (ica_get_functionlist(pmech_list, &mech_len) != 0) {
		perror("get_functionlist: ");
		free(pmech_list);
		return 0;
	}

	for (j = 0; j < mech_len; j++) {
		if (pmech_list[j].mech_mode_id == EC_DSA_SIGN) {
			if (pmech_list[j].flags & ICA_FLAG_DHW) {
				free(pmech_list);
				return 1;
			}
		}
	}

	free(pmech_list);
	return 0;
}

int rsa_keylen_supported_by_openssl(unsigned int modulus_bitlength)
{
	unsigned char modexpo_public_e[512] = { 0 };
	unsigned char modexpo_public_n[512] = { 0 };
	unsigned char crt_private_p[256] = { 0 };
	unsigned char crt_private_q[256] = { 0 };
	unsigned char crt_private_dp[256] = { 0 };
	unsigned char crt_private_dq[256] = { 0 };
	unsigned char crt_private_inv_q[256] = { 0 };
	ica_adapter_handle_t ah;
	ica_rsa_key_mod_expo_t public_key;
	ica_rsa_key_crt_t private_key;
	int rc;

	rc = ica_open_adapter(&ah);
	if (rc != 0)
		return 0;

	public_key.modulus = modexpo_public_n;
	public_key.exponent = modexpo_public_e;
	public_key.key_length = (modulus_bitlength + 7) / 8;

	private_key.p = crt_private_p;
	private_key.q = crt_private_q;
	private_key.dp = crt_private_dp;
	private_key.dq = crt_private_dq;
	private_key.qInverse = crt_private_inv_q;
	private_key.key_length = (modulus_bitlength + 7) / 8;

	rc = ica_rsa_key_generate_crt(ah, modulus_bitlength,
							&public_key, &private_key);

	ica_close_adapter(ah);

	return rc == 0 ? 1 : 0;
}

int get_rsa_minlen(void)
{
	int keylen_array[] = { 57, 512, 1024, 2048, 4096 };
	size_t i;

	for (i = 0; i < sizeof(keylen_array) / sizeof(int); i++) {
		if (rsa_keylen_supported_by_openssl(keylen_array[i])) {
			return keylen_array[i];
		}
	}

	return -1;
}

/**
 * Print out the minimum and maximum RSA key length. The maximum length is
 * restricted to 4096 bits by crypto cards. The minimum accepted length in
 * libica is 57 bits, but the available min length depends on the openssl
 * version and fips mode.
 */
void print_rsa(void)
{
	int minlen = get_rsa_minlen();

	if (minlen > 0)
		printf("RSA key lengths: %d ... 4096 bits.\n", minlen);
	else
		printf("Error: cannot determine supported RSA key lengths via openssl.\n");

#ifdef ICA_FIPS
	printf("Built-in FIPS support: FIPS 140-3 mode %s.\n",
	    ica_fips_status() & ICA_FIPS_MODE ? "active" : "inactive");
	if (ica_fips_status() >> 1)
		printf("FIPS SELF-TEST FAILURE. CHECK THE SYSLOG.\n");
#else
	printf("No built-in FIPS support.\n");
#endif /* ICA_FIPS */
}

static int num_cpacf_curves(void)
{
	if (!is_msa9())
		return 0;

	return sizeof(cpacf_nids) / sizeof(unsigned int);
}

static int num_cca_curves(void)
{
	if (!online_cca_card())
		return 0;

	return sizeof(cca_nids) / sizeof(unsigned int);
}

void append_nid(unsigned int nid, unsigned int flag,
				s390_supported_curves_t *curve_array, size_t array_len)
{
	const char* sname;
	unsigned int i;

	for (i = 0; i < array_len; i++) {
		if (curve_array[i].nid == 0) {
			curve_array[i].nid = nid;
			sname = OBJ_nid2sn(nid);
			if (sname == NULL)
				sname = "";
			curve_array[i].sname = (unsigned char *)sname;
			curve_array[i].flags |= flag;
#ifndef NO_SW_FALLBACKS
			if (curve_supported_via_openssl(nid) && sw_fallbacks_implemented(nid))
				curve_array[i].flags |= ICA_FLAG_SW;
#endif
			break;
		}
	}
}

void add_hw_curves(s390_supported_curves_t *curve_array, size_t array_len,
				 const unsigned int nids[], unsigned int nids_len,
				 unsigned int flag)
{
	unsigned int i, j, merged;

	for (i = 0; i < nids_len; i++) {
		merged = 0;
		for (j = 0; j < array_len; j++) {
			if (curve_array[j].nid == nids[i]) {
				curve_array[j].flags |= flag;
#ifndef NO_SW_FALLBACKS
				if (curve_supported_via_openssl(nids[i]) && sw_fallbacks_implemented(nids[i]))
					curve_array[j].flags |= ICA_FLAG_SW;
#endif
				merged = 1;
				break;
			}
		}
		if (!merged)
			append_nid(nids[i], flag, curve_array, array_len);
	}
}

void add_curves(s390_supported_curves_t *curve_array, size_t array_len)
{
#ifdef ICA_FIPS
	if (is_msa9()) {
		if (ica_fips_status() & ICA_FIPS_MODE) {
			add_hw_curves(curve_array, array_len, cpacf_fips_nids, cpacf_fips_nids_len, ICA_FLAG_SHW);
		} else {
			add_hw_curves(curve_array, array_len, cpacf_nids, cpacf_nids_len, ICA_FLAG_SHW);
			if (online_cca_card())
				add_hw_curves(curve_array, array_len, cca_nids, cca_nids_len, ICA_FLAG_DHW);
		}
	} else {
		/* No MSA9, ECC only available via CEX cards */
		if (online_cca_card())
			add_hw_curves(curve_array, array_len, cca_nids, cca_nids_len, ICA_FLAG_DHW);
	}
#else
	if (online_cca_card())
		add_hw_curves(curve_array, array_len, cca_nids, cca_nids_len, ICA_FLAG_DHW);

	if (is_msa9())
		add_hw_curves(curve_array, array_len, cpacf_nids, cpacf_nids_len, ICA_FLAG_SHW);
#endif
}

void print_ec_curves(void)
{
	s390_supported_curves_t *curve_array;
	unsigned int array_len, array_size, n;
#ifdef NO_CPACF
	char *no_shw = "-";
#else
	char *no_shw = "no";
#endif
#ifdef NO_SW_FALLBACKS
	char *no_sw = "-";
#else
	char *no_sw = "no";
#endif

	array_len = num_cca_curves() + num_cpacf_curves();
	array_size = array_len * sizeof(s390_supported_curves_t);
	curve_array = calloc(1, array_size);
	if (!curve_array) {
		fprintf(stderr, "Error: cannot allocate %d bytes for array of curves.\n", array_size);
		return;
	}

	add_curves(curve_array, array_len);

	printf("-------------------------------------------------------\n");
	printf("                 |         hardware        |           \n");
	printf("        EC curve |   dynamic  |   static   |  software \n");
	printf("                 |            |  (msa=%02d)  |         \n",ica_get_msa_level());
	printf("-----------------+------------+------------+-----------\n");

	for (n = 0; n < array_len; n++) {
		if (curve_array[n].nid != 0) {
			printf("%16s |    %*s     |    %*s     |    %*s\n",
				curve_array[n].sname,
				CELL_SIZE,
				curve_array[n].flags & ICA_FLAG_DHW ? "yes" : "no",
				CELL_SIZE,
				curve_array[n].flags & ICA_FLAG_SHW ? "yes" : no_shw,
				CELL_SIZE,
				curve_array[n].flags & ICA_FLAG_SW ? "yes" : no_sw);
		}
	}
	printf("-------------------------------------------------------\n");
#ifdef ICA_FIPS
	printf("Built-in FIPS support: FIPS 140-3 mode %s.\n",
	    ica_fips_status() & ICA_FIPS_MODE ? "active" : "inactive");
	if (ica_fips_status() >> 1)
		printf("FIPS SELF-TEST FAILURE. CHECK THE SYSLOG.\n");
#else
	printf("No built-in FIPS support.\n");
#endif /* ICA_FIPS */

	free(curve_array);
}

void print_version(void)
{
#if defined(NO_SW_FALLBACKS) && defined(NO_CPACF)
	printf(CMD_NAME ": libica-cex version " VERSION "\n" COPYRIGHT "\n");
#else
	printf(CMD_NAME ": libica version " VERSION "\n" COPYRIGHT "\n");
#endif
}

void print_help(char *cmd)
{
	printf("Usage: %s [OPTION]\n", cmd);
	printf
	    ("Display a list of all CP Assist for Cryptographic Function "
	     "(CPACF)\noperations supported by libica on this system.\n"
	     "\n" "Options:\n"
	     " -v, --version        show version information\n"
	     " -c, --list-curves    list supported EC curves\n"
	     " -r, --list-rsa       list supported RSA key lengths\n"
	     " -h, --help           display this help text\n");
}

#define getopt_string "qcrvh"
static struct option getopt_long_options[] = {
	{"list-curves", 0, 0, 'c'},
	{"list-rsa", 0, 0, 'r'},
	{"version", 0, 0, 'v'},
	{"help", 0, 0, 'h'},
	{0, 0, 0, 0}
};


struct crypt_pair {
	char *name;
	unsigned int algo_id;
};

static struct crypt_pair crypt_map[] = {
	{"SHA-1", SHA1},
	{"SHA-224", SHA224},
	{"SHA-256", SHA256},
	{"SHA-384", SHA384},
	{"SHA-512", SHA512},
	{"SHA-512/224", SHA512_224},
	{"SHA-512/256", SHA512_256},
	{"SHA3-224", SHA3_224},
	{"SHA3-256", SHA3_256},
	{"SHA3-384", SHA3_384},
	{"SHA3-512", SHA3_512},
	{"SHAKE-128", SHAKE128},
	{"SHAKE-256", SHAKE128},
	{"GHASH", G_HASH},
	{"P_RNG", P_RNG},
	{"DRBG-SHA-512", SHA512_DRNG},
	{"ECDH", EC_DH},
	{"ECDSA Sign", EC_DSA_SIGN},
	{"ECDSA Verify", EC_DSA_VERIFY},
	{"EC Keygen", EC_KGEN},
	{"Ed25519 Keygen", ED25519_KEYGEN},
	{"Ed25519 Sign", ED25519_SIGN},
	{"Ed25519 Verify", ED25519_VERIFY},
	{"Ed448 Keygen", ED448_KEYGEN},
	{"Ed448 Sign", ED448_SIGN},
	{"Ed448 Verify", ED448_VERIFY},
	{"X25519 Keygen", X25519_KEYGEN},
	{"X25519 Derive", X25519_DERIVE},
	{"X448 Keygen", X448_KEYGEN},
	{"X448 Derive", X448_DERIVE},
	{"RSA Keygen ME", RSA_KEY_GEN_ME},
	{"RSA Keygen CRT", RSA_KEY_GEN_CRT},
	{"RSA ME", RSA_ME},
	{"RSA CRT", RSA_CRT},
	{"DES ECB", DES_ECB},
	{"DES CBC", DES_CBC},
	{"DES OFB", DES_OFB},
	{"DES CFB", DES_CFB},
	{"DES CTR", DES_CTR},
	{"DES CMAC", DES_CMAC},
	{"3DES ECB", DES3_ECB},
	{"3DES CBC", DES3_CBC},
	{"3DES OFB", DES3_OFB},
	{"3DES CFB", DES3_OFB},
	{"3DES CTR", DES3_CTR},
	{"3DES CMAC", DES3_CMAC},
	{"AES ECB", AES_ECB},
	{"AES CBC", AES_CBC},
	{"AES OFB", AES_OFB},
	{"AES CFB", AES_CFB},
	{"AES CTR", AES_CTR},
	{"AES CMAC", AES_CMAC},
	{"AES XTS", AES_XTS},
	{"AES GCM", AES_GCM_KMA},
	{NULL,0}
};


int main(int argc, char **argv)
{
	int rc;
	int index = 0;
	unsigned int mech_len, j;
	libica_func_list_element *pmech_list = NULL;
	unsigned int i;
	char *no_dhw = "no";
#ifdef NO_CPACF
	char *no_shw = "-";
#else
	char *no_shw = "no";
#endif
#ifdef NO_SW_FALLBACKS
	char *no_sw = "-";
#else
	char *no_sw = "no";
#endif

	while ((rc = getopt_long(argc, argv, getopt_string,
				 getopt_long_options, &index)) != -1) {
		switch (rc) {
		case 'c':
			print_ec_curves();
			exit(0);
		case 'r':
			print_rsa();
			exit(0);
		case 'v':
			print_version();
			exit(0);
			break;
		case 'h':
			print_help(basename(argv[0]));
			exit(0);
		default:
			fprintf(stderr, "Try '%s --help' for more"
				" information.\n", basename(argv[0]));
			ica_cleanup();
			exit(1);
		}
	}
	if (optind < argc) {
		fprintf(stderr, "%s: invalid option.\n"
			"Try '%s --help' for more information.\n",
			argv[0], basename(argv[0]));
		ica_cleanup();
		exit(1);
	}

	printf("              Cryptographic algorithm support      \n");
	printf("------------------------------------------------------\n");

	if (ica_get_functionlist(NULL, &mech_len) != 0){
		perror("get_functionlist: ");
		ica_cleanup();
		return EXIT_FAILURE;
	}
	pmech_list = malloc(sizeof(libica_func_list_element)*mech_len);
	if (ica_get_functionlist(pmech_list, &mech_len) != 0){
		perror("get_functionlist: ");
		free(pmech_list);
		ica_cleanup();
		return EXIT_FAILURE;
	}

	printf("               |         hardware        |            \n");
	printf(" function      |   dynamic  |   static   |  software  \n");
	printf("               |            |  (msa=%02d)  |            \n",ica_get_msa_level());
	printf("---------------+------------+------------+------------\n");
	for (i = 0; crypt_map[i].algo_id; i++) {
		for (j = 0; j < mech_len; j++) {
			if (crypt_map[i].algo_id == pmech_list[j].mech_mode_id) {
#ifdef ICA_FIPS
				if (((ica_fips_status() & ICA_FIPS_MODE) &&
					!fips_approved(pmech_list[j].mech_mode_id)) ||
					ica_fips_status() >> 1) {
					printf("%14s |  blocked   |  blocked   |  blocked\n",
						crypt_map[i].name);
					break;
				}
#endif /* ICA_FIPS */
				printf("%14s |    %*s     |    %*s     |    %*s     \n",
					crypt_map[i].name,
					CELL_SIZE,
					pmech_list[j].flags & ICA_FLAG_DHW ? "yes" : no_dhw,
					CELL_SIZE,
					pmech_list[j].flags & ICA_FLAG_SHW ? "yes" : no_shw,
					CELL_SIZE,
					pmech_list[j].flags & ICA_FLAG_SW ? "yes" : no_sw);
			}
		}
	}
	free(pmech_list);
	printf("------------------------------------------------------\n");

#ifdef ICA_FIPS
	printf("Built-in FIPS support: FIPS 140-3 mode %s.\n",
		ica_fips_status() & ICA_FIPS_MODE ? "active" : "inactive");
	if (ica_fips_status() >> 1)
		printf("FIPS SELF-TEST FAILURE. CHECK THE SYSLOG.\n");
#else
	printf("No built-in FIPS support.\n");
#endif /* ICA_FIPS */

#ifdef NO_SW_FALLBACKS
	printf("Software fallbacks are disabled.\n");
#endif
#ifdef NO_CPACF
	printf("CPACF support (including fallbacks) is disabled in libica-cex.\n");
#endif

	ica_cleanup();
	return EXIT_SUCCESS;
}
